generic3g/MAPL_Generic.F90

```diff
@@ -575,26 +598,85 @@ subroutine gridcomp_connect_all(gridcomp, src_comp, dst_comp, rc)
       _RETURN(_SUCCESS)
    end subroutine gridcomp_connect_all
 
-   subroutine hconfig_get_string(hconfig, keystring, value, default, rc)
-      type(ESMF_HConfig), intent(inout) :: hconfig
-      character(*), intent(in) :: keystring
-      character(:), allocatable :: value
-      character(*), optional, intent(in) :: default
+   subroutine gridcomp_get_hconfig(gridcomp, hconfig, rc)
+      type(ESMF_GridComp), intent(inout) :: gridcomp
+      type(ESMF_HConfig), intent(out) :: hconfig
       integer, optional, intent(out) :: rc
 
       integer :: status
-      logical :: has_key
+      type(ESMF_Config) :: config
+      
+      call ESMF_GridCompGet(gridcomp, config=config, _RC)
+      call ESMF_ConfigGet(config, hconfig=hconfig, _RC)
       
-      has_key = ESMF_HConfigIsDefined(hconfig, keystring=keystring, _RC)
-      if (has_key) then
-         value = ESMF_HConfigAsSTring(hconfig, keystring=keystring, _RC)
-         _RETURN(_SUCCESS)
+
+      _RETURN(_SUCCESS)
+   end subroutine gridcomp_get_hconfig
+
+   subroutine mapl_resource_get_scalar(hconfig, keystring, value, unusable, default, logger, is_default, found, rc)
```

Comment:
I thought this procedure would go in the new directory you created.  Please explain your thinking.

Reply:
My plan was to separate the `ESMF_HConfig` handling from the logger and default handling. `hconfig_utils` uses parts of `ESMF`. `mapl3hconfig_get` (public) uses `ESMF_HConfig`. `mapl3hconfig_get_private` uses `ESMF_HConfig`, some `ESMF_HConfig` procedures, and ESMF kinds. It passes up message strings up instead of directly accessing the logger. Since there is no logger associated with the hconfig, I decided not to pass it in, because the calling procedure would have to get the logger to pass in. Similarly, there is no default value for a keystring in the hconfig, so I decided to let the calling procedure handle the default. I thought about throwing an exception **if** a value for the keystring is not found **and** no default is found, but I decided that it wasn't a true error condition. So, I added a flag to signal that the value was not found, and let the calling procedure use that flag to handle that case.